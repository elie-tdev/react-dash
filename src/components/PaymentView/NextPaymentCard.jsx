import React, { useState } from 'react';
import cn from 'classnames';
import { Box, Card, CardContent, makeStyles, Typography, Divider } from '@material-ui/core';
import CalendarIcon from '@material-ui/icons/CalendarToday';
import { useTranslation } from 'react-i18next';

import { LabelValueColumn, PrimaryButton, SecondaryButton, SubmitPaymentModal } from '@/components';
import useGlobalStyles from '@/themes/useGlobalStyles';

const useStyles = makeStyles(theme => ({
  icon: {
    fontSize: '14px',
    marginRight: '8px',
    color: theme.palette.primary.main,
  },
  box: {
    padding: '16px',
  },
  row: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'baseline',
    '&:not(:last-child)': {
      marginBottom: '16px',
    },
  },
  headerRow: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'baseline',
  },
  pastDueText: {
    paddingTop: '0px',
  },
}));

const NextPaymentCard = ({
  nextPaymentDays,
  paymentAmount,
  dueDate,
  elevation = 2,
  isOutlined,
  isCurrent,
  isAutoPay,
  onSchedulePayment,
  onEnrollInAutoPay,
  isLoanOverviewScreen = false,
}) => {
  const globalStyles = useGlobalStyles();
  const styles = useStyles();
  const { t } = useTranslation();

  const [isSubmitPaymentOpen, setIsSubmitPaymentOpen] = useState(false);
  const [isAutoPaySetup, setIsAutoPaySetup] = useState(false);

  nextPaymentDays = nextPaymentDays.includes('hours') ? 'in 1 day' : nextPaymentDays;

  const cardProps = {};
  if (isOutlined) {
    cardProps.variant = 'outlined';
  }
  return (
    <Card
      data-testid="NextPayment"
      className={cn(globalStyles.card)}
      elevation={elevation}
      {...cardProps}
    >
      <CardContent>
        <Card className={cn(globalStyles.cardBorder, globalStyles.cardBackground)} elevation={0}>
          <Box className={cn(styles.box)}>
            <Box className={cn(styles.headerRow)}>
              <Typography variant="h5">
                <CalendarIcon className={cn(styles.icon)} />
                {t('content.Next Payment')}
              </Typography>
              <Box className={cn(globalStyles.due)}>
                {t('content.Due')} {nextPaymentDays}
              </Box>
            </Box>
          </Box>
        </Card>
        <Box className={cn(styles.box, styles.noBottomPadding)}>
          <Box className={cn(styles.row)}>
            <LabelValueColumn
              variant={isCurrent ? 'primary' : null}
              size="lg"
              label={t('content.Payment Amount')}
              value={paymentAmount}
            />
            <LabelValueColumn size="lg" label={t('content.Due Date')} value={dueDate} />
          </Box>
        </Box>
      </CardContent>
      {isCurrent ? (
        <>
          <Divider />
          <CardContent className={cn(globalStyles.cardContent)} elevation={0}>
            <Box className={cn(globalStyles.column)}>
              <Box className={cn(globalStyles.column, styles.row)}>
                <Typography
                  className={cn(styles.infoTitle, globalStyles.importantText)}
                  variant="h5"
                >
                  <CalendarIcon className={cn(styles.icon)} />
                  {t('content.Payment Due')}
                </Typography>
                <Typography className={cn(globalStyles.caption)}>
                  {t('content.Payment Due Body', {
                    dueDate: dueDate,
                  })}
                </Typography>
              </Box>
              {isAutoPay || isLoanOverviewScreen ? (
                <Box className={cn(styles.row)}>
                  <PrimaryButton onClick={() => setIsSubmitPaymentOpen(true)}>
                    {t('content.Schedule a Payment')}
                  </PrimaryButton>
                </Box>
              ) : (
                <>
                  <Box className={cn(styles.row)}>
                    <SecondaryButton onClick={() => setIsSubmitPaymentOpen(true)}>
                      {t('content.Schedule a Payment')}
                    </SecondaryButton>
                  </Box>
                  <Box className={cn(styles.row)}>
                    <PrimaryButton onClick={() => setIsAutoPaySetup(true)}>
                      {t('content.Enroll in AutoPay')}
                    </PrimaryButton>
                  </Box>
                </>
              )}
            </Box>
          </CardContent>
        </>
      ) : (
        <CardContent className={cn(globalStyles.cardContent, styles.pastDueText)}>
          <Box className={cn(globalStyles.column)}>
            <Box className={cn(styles.row)}>
              <Typography className={cn(globalStyles.caption)}>
                {t('content.Payment Past Due Body')}
              </Typography>
            </Box>
          </Box>
        </CardContent>
      )}

      {(isSubmitPaymentOpen || isAutoPaySetup) && (
        <SubmitPaymentModal
          open={isSubmitPaymentOpen || isAutoPaySetup}
          isAutoPaySetup={isAutoPaySetup}
          onClose={() => {
            setIsSubmitPaymentOpen(false);
            setIsAutoPaySetup(false);
          }}
          onSuccess={isAutoPaySetup ? onEnrollInAutoPay : onSchedulePayment}
        />
      )}
    </Card>
  );
};

export default NextPaymentCard;
