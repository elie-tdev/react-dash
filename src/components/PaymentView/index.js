export { default as PaymentView } from './PaymentView';
export { default as PaymentCurrent } from './PaymentCurrent';
export { default as PaymentPastDue } from './PaymentPastDue';
export { default as ScheduledPayment } from './ScheduledPayment';
export { default as NextPaymentView } from './NextPaymentView';
export { default as EnrollInAutoPay } from './EnrollInAutoPay';
export { default as NextPaymentCard } from './NextPaymentCard';
export { default as PastDueCard } from './PastDueCard';
export { default as AutoPayCard } from './AutoPayCard';
