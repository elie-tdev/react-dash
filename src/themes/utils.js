export const opacity = (hex, opacity) => {
  var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  if (!result) {
    throw new Error('Invalid hex', hex);
  }
  return `rgba(${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(
    result[3],
    16,
  )}, ${opacity})`;
};

export const gradient = ({ type = 'linear', colors, background }) =>
  `${type}-gradient(${colors.join(', ')}), ${background};`;
