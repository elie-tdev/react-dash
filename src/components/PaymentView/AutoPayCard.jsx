import React, { useState } from 'react';
import cn from 'classnames';
import {
  Box,
  Card,
  CardActions,
  CardContent,
  Divider,
  makeStyles,
  Typography,
} from '@material-ui/core';
import Autorenew from '@material-ui/icons/Autorenew';
import { useTranslation } from 'react-i18next';

import PaymentInfo from './PaymentInfo';
import PaymentLabelValue from './PaymentLabelValue';

import useGlobalStyles from '@/themes/useGlobalStyles';
import { PrimaryButton, SubmitPaymentModal } from '@/components';
import { gradient } from '@/themes';

const useStyles = makeStyles(theme => ({
  rootDark: {
    background: gradient({
      type: 'radial',
      colors: theme.palette.common.gradient.primary.dark,
      background: theme.palette.primary.main,
    }),
  },
  contractText: {
    color: theme.palette.primary.contrastText,
  },
  headerRow: {
    marginBottom: '16px',
  },
  icon: {
    marginLeft: '0',
  },
  divider: {
    backgroundColor: 'rgba(255, 255, 255, 0.1)',
  },
  automaticPaymentTitle: {
    marginBottom: '8px',
  },
  iconLabel: {
    display: 'flex',
    flex: 1,
    color: theme.palette.primary.main,
    alignItems: 'center',
  },
  autoRenewIcon: {
    height: '32px',
    width: '32px',
    padding: '8px',
    borderRadius: '40px',
    backgroundColor: theme.palette.primary.light,
    color: theme.palette.primary.contrastText,
  },
}));

const AutoPayCard = ({
  paymentDate,
  paymentValue,
  nextPaymentDays,
  onManagePayment,
  _autoPayRecord,
}) => {
  const globalStyles = useGlobalStyles();
  const styles = useStyles();
  const { t } = useTranslation();
  const [isAutoPayManage, setIsAutoPayManage] = useState(false);

  nextPaymentDays = nextPaymentDays.includes('hours') ? 'in 1 day' : nextPaymentDays;

  return (
    <Card
      data-testid="AutoPayCard"
      className={cn(globalStyles.card, styles.rootDark)}
      elevation={2}
    >
      <CardContent className={cn(globalStyles.cardContent)}>
        <Box className={cn(globalStyles.row, styles.headerRow)}>
          <Box className={cn(styles.iconLabel)}>
            <Box className={cn(styles.iconLabel)}>
              <Autorenew className={cn(globalStyles.icon, styles.autoRenewIcon)} />
              <Typography className={cn(styles.contractText)} variant="body2" noWrap>
                {t('content.AutoPay')}
              </Typography>
            </Box>
            <Typography className={cn(styles.contractText)} variant="body2" noWrap>
              {t('content.Scheduled Payment Sent', {
                days: nextPaymentDays,
              })}
            </Typography>
          </Box>
        </Box>
        <Box className={cn(globalStyles.row)}>
          <PaymentLabelValue
            paymentType="autoPay"
            paymentValue={paymentValue}
            paymentDate={paymentDate}
            dark
          />
        </Box>
        <Divider className={cn(globalStyles.divider, globalStyles.dividerWide, styles.divider)} />
        <Box className={cn(globalStyles.row, styles.automaticPaymentTitle)}>
          <Box>
            <Autorenew className={cn(globalStyles.icon, styles.contractText, styles.icon)} />
          </Box>
          <Box>
            <Typography className={cn(styles.contractText)} variant="h5">
              {t('content.Automatic Payment')}
            </Typography>
          </Box>
        </Box>
        <PaymentInfo
          paymentAmount={paymentValue}
          paymentDate={paymentDate}
          dark
          paymentType="autoPay"
        />
      </CardContent>
      <CardActions className={cn(globalStyles.cardActions)}>
        <PrimaryButton
          onClick={() => {
            setIsAutoPayManage(true);
          }}
          rootStyles={{ backgroundColor: 'rgba(255, 255, 255, 0.1)' }}
        >
          {t('content.Manage AutoPay')}
        </PrimaryButton>
      </CardActions>
      {isAutoPayManage && (
        <SubmitPaymentModal
          isManagePayment={true}
          open={isAutoPayManage}
          onClose={() => {
            setIsAutoPayManage(false);
          }}
          onSuccess={onManagePayment}
        />
      )}
    </Card>
  );
};

export default AutoPayCard;
