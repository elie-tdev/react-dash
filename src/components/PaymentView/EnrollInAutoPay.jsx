import cn from 'classnames';
import React, { useState } from 'react';
import { Card, makeStyles, Typography, Box } from '@material-ui/core';
import { useTranslation } from 'react-i18next';

import autopayCardBackground from '@/static/images/autopayCardBackground.svg';
import { gradient, useGlobalStyles } from '@/themes';
import { PrimaryButton, SubmitPaymentModal } from '@/components';

const useStyles = makeStyles(theme => ({
  card: {
    minHeight: '240px',
    background: gradient({
      type: 'radial',
      colors: theme.palette.common.gradient.primary.dark,
      background: theme.palette.primary.main,
    }),
  },
  box: {
    display: 'flex',
    justifyContent: 'space-between',
    flexDirection: 'column',
    alignItems: 'flex-start',
    padding: '24px',
    height: '100%',
    minHeight: 'inherit',
    backgroundImage: `url(${autopayCardBackground})`,
    backgroundRepeat: 'no-repeat',
    backgroundPosition: 'center center',
    backgroundSize: '100% 100%',
  },
  title: {
    width: '75%',
    fontWeight: 600,
    fontSize: '22px',
    lineHeight: '32px',
    color: theme.palette.primary.contrastText,
    [theme.breakpoints.down('md')]: {
      width: '100%',
    },
  },
  button: {
    height: 'auto',
  },
}));

const EnrollInAutoPay = ({ onEnrollInAutoPay }) => {
  const { t } = useTranslation();
  const styles = useStyles();
  const globalStyles = useGlobalStyles();
  const [isAutoPaySetup, setIsAutoPaySetup] = useState(false);

  return (
    <Card className={cn(globalStyles.card, styles.card)}>
      <Box className={cn(styles.box)}>
        <Typography variant="h1" className={styles.title}>
          {t('content.Enroll in AutoPay and never miss a payment again')}
        </Typography>
        <PrimaryButton
          variant="text"
          fullWidth={false}
          onClick={() => setIsAutoPaySetup(true)}
          className={styles.button}
        >
          {t('content.Enroll in AutoPay')}
        </PrimaryButton>
      </Box>
      {isAutoPaySetup && (
        <SubmitPaymentModal
          open={isAutoPaySetup}
          isAutoPaySetup={isAutoPaySetup}
          onClose={() => {
            setIsAutoPaySetup(false);
          }}
          onSuccess={onEnrollInAutoPay}
        />
      )}
    </Card>
  );
};

export default EnrollInAutoPay;
