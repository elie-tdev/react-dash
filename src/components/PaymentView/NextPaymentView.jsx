import { useMemo } from 'react';
import dateformat from 'dateformat';
import { useParams } from 'react-router';
import TimeAgo from 'javascript-time-ago';

import { useLoanState, useOneTimePayments } from '@/customHooks';
import { formatDollar } from '@/utils/format';
import { formatDateString } from '@/properties';
import { ScheduledPayment, AutoPayCard, NextPaymentCard } from '@/components';

const NextPaymentView = ({
  loanNumberProp,
  isPaymentHistory = false,
  loanDataNextPayment,
  isLoanOverviewScreen,
  onSchedulePayment,
  onEnrollInAutoPay,
  onManagePayment,
}) => {
  const { loanNumber } = loanNumberProp ? loanNumberProp : useParams();

  const {
    isAutoPay,
    isCurrent,
    isBilled,
    hasOneTimePayment,
    oneTimePaymentRowIds,
    nextPayment,
    autoPayRecord,
    achLoanNumber,
  } = useLoanState(loanNumber);
  const { oneTimePayments } = useOneTimePayments(oneTimePaymentRowIds);

  const nextPaymentData = useMemo(() => {
    if (nextPayment.Date_Due) {
      const nextPaymentDays = new TimeAgo('en-US').format(nextPayment.Date_Due);
      const paymentAmount = formatDollar(nextPayment.Payment_Amount);
      const dueDate = dateformat(nextPayment.Date_Due, formatDateString);
      return { nextPaymentDays, paymentAmount, dueDate };
    }
  }, [nextPayment]);

  const isBetweenDays = (scheduledDate, daysNum) => {
    const timeDiff = new Date(scheduledDate) - new Date();

    return timeDiff <= daysNum * 24 * 60 * 60 * 1000 && timeDiff >= 0;
  };

  const oneTimePaymentsData = useMemo(() => {
    if (!hasOneTimePayment || oneTimePayments.error || !oneTimePayments.activePayments.length) {
      return;
    }

    const futureOneTimePayments = oneTimePayments.activePayments.filter(payment => {
      return new Date(payment.scheduledDate) - new Date() >= 1;
    });

    const activeOneTimePayments =
      futureOneTimePayments && futureOneTimePayments.length > 1
        ? futureOneTimePayments.sort(
            (a, b) => new Date(a.scheduledDate) - new Date(b.scheduledDate),
          )
        : futureOneTimePayments;

    const pendingPayments =
      activeOneTimePayments &&
      activeOneTimePayments.length >= 1 &&
      activeOneTimePayments.filter(payment => {
        return isBetweenDays(payment.scheduledDate, 2);
      });

    const isOneTimePaymentPending =
      activeOneTimePayments &&
      activeOneTimePayments.length >= 1 &&
      isBetweenDays(activeOneTimePayments[0].scheduledDate, 2);

    const activePendingPayments =
      pendingPayments &&
      pendingPayments.length > 0 &&
      pendingPayments.map(payment => {
        return {
          scheduledPaymentDays: new TimeAgo('en-US').format(
            new Date(payment.scheduledDate).getTime() +
              new Date(payment.scheduledDate).getTimezoneOffset() * 60000 * 4,
          ),
          scheduledDate: dateformat(new Date(payment.scheduledDate), formatDateString),
          amount: payment.amount,
          accountNumber: payment.accountNumber,
          isViewOnly: isBetweenDays(payment.scheduledDate, 2),
          paymentRowId: payment.rowId,
        };
      });
    const activePaymentsNum = activeOneTimePayments.length ?? 0;
    const pendingPaymentsNum = activePendingPayments.length ?? 0;

    const pending = isOneTimePaymentPending
      ? {
          scheduledPaymentDays: new TimeAgo('en-US').format(
            new Date(activeOneTimePayments[0].scheduledDate).getTime() +
              new Date(activeOneTimePayments[0].scheduledDate).getTimezoneOffset() * 60000 * 4,
          ),
          scheduledDate: dateformat(
            new Date(activeOneTimePayments[0].scheduledDate),
            formatDateString,
          ),
          amount: activeOneTimePayments[0].amount,
          accountNumber: activeOneTimePayments[0].accountNumber,
          isViewOnly: true,
          paymentRowId: activeOneTimePayments[0].rowId,
        }
      : {};

    let scheduled;
    if (
      activeOneTimePayments &&
      activeOneTimePayments.length >= 1 &&
      activePaymentsNum - pendingPaymentsNum >= 1
    ) {
      scheduled = {
        scheduledPaymentDays: new TimeAgo('en-US').format(
          new Date(activeOneTimePayments[pendingPaymentsNum].scheduledDate).getTime() +
            new Date(activeOneTimePayments[pendingPaymentsNum].scheduledDate).getTimezoneOffset() *
              60000 *
              4,
        ),
        scheduledDate: dateformat(
          new Date(activeOneTimePayments[pendingPaymentsNum].scheduledDate),
          formatDateString,
        ),
        amount: activeOneTimePayments[pendingPaymentsNum].amount,
        accountNumber: activeOneTimePayments[pendingPaymentsNum].accountNumber,
        isViewOnly: isBetweenDays(activeOneTimePayments[pendingPaymentsNum].scheduledDate, 2),
        paymentRowId: activeOneTimePayments[pendingPaymentsNum].rowId,
      };
    }
    const otherOneTimePaymentsNum =
      activeOneTimePayments.length - (pendingPaymentsNum > 0 ? pendingPaymentsNum + 1 : 1);

    const otherOneTimePayments =
      otherOneTimePaymentsNum >= 1 &&
      activeOneTimePayments &&
      activeOneTimePayments.length > 1 &&
      activeOneTimePayments.slice(-otherOneTimePaymentsNum).map(payment => {
        return {
          scheduledPaymentDays: new TimeAgo('en-US').format(
            new Date(payment.scheduledDate).getTime() +
              new Date(payment.scheduledDate).getTimezoneOffset() * 60000 * 4,
          ),
          scheduledDate: dateformat(new Date(payment.scheduledDate), formatDateString),
          amount: payment.amount,
          accountNumber: payment.accountNumber,
          isViewOnly: isBetweenDays(payment.scheduledDate, 2),
          paymentRowId: payment.rowId,
        };
      });

    return {
      isOneTimePaymentPending,
      otherOneTimePaymentsNum,
      otherOneTimePayments,
      pending,
      scheduled,
      activePendingPayments,
    };
  }, [hasOneTimePayment, oneTimePayments]);

  const autoPayData = useMemo(() => {
    if (!loanDataNextPayment || !isAutoPay) {
      return;
    }
    let paymentDate;
    if (!isBilled) {
      paymentDate =
        loanDataNextPayment.Loans_GetPaymentInfo.payload.data.Next_Principal_Payment_Date;
    } else {
      paymentDate =
        loanDataNextPayment.Loans_GetPaymentInfo.payload.data.Current_Principal_Payment_Date;
    }
    return {
      paymentAmount: loanDataNextPayment.Loans_GetPaymentInfo.payload.data.Amortized_Payment_Amount,
      paymentDate: dateformat(new Date(paymentDate), formatDateString),
      nextPaymentDays: new TimeAgo('en-US').format(new Date(paymentDate)),
    };
  }, [loanDataNextPayment, isBilled, isCurrent, isAutoPay]);

  return (
    <>
      {isAutoPay && autoPayData && (
        <AutoPayCard
          paymentType="autoPay"
          paymentDate={autoPayData.paymentDate}
          paymentValue={autoPayData.paymentAmount}
          nextPaymentDays={autoPayData.nextPaymentDays}
          dark
          autoPayRecord={autoPayRecord}
        />
      )}
      {oneTimePaymentsData &&
        oneTimePaymentsData?.activePendingPayments &&
        oneTimePaymentsData?.activePendingPayments.length > 0 &&
        oneTimePaymentsData?.activePendingPayments.map((pendingPayment, index) => (
          <ScheduledPayment
            key={index}
            scheduledPaymentDate={pendingPayment.scheduledDate}
            paymentAmount={pendingPayment.amount}
            scheduledPaymentDays={pendingPayment.scheduledPaymentDays}
            isPending
            isViewOnly={pendingPayment.isViewOnly}
            isPaymentHistory={isPaymentHistory}
            accountNumber={pendingPayment.accountNumber}
            paymentRowId={pendingPayment.paymentRowId}
            achLoanNumber={achLoanNumber}
            onManagePayment={onManagePayment}
          />
        ))}
      {oneTimePaymentsData && oneTimePaymentsData?.scheduled && (
        <ScheduledPayment
          scheduledPaymentDays={oneTimePaymentsData.scheduled.scheduledPaymentDays}
          scheduledPaymentDate={oneTimePaymentsData.scheduled.scheduledDate}
          paymentAmount={oneTimePaymentsData.scheduled.amount}
          otherOneTimePaymentsNum={oneTimePaymentsData.otherOneTimePaymentsNum}
          isViewOnly={oneTimePaymentsData.scheduled.isViewOnly}
          isPaymentHistory={isPaymentHistory}
          accountNumber={oneTimePaymentsData.scheduled.accountNumber}
          paymentRowId={oneTimePaymentsData.scheduled.paymentRowId}
          achLoanNumber={achLoanNumber}
          onManagePayment={onManagePayment}
        />
      )}
      {oneTimePaymentsData &&
        oneTimePaymentsData?.otherOneTimePayments.length >= 1 &&
        isPaymentHistory &&
        oneTimePaymentsData?.otherOneTimePaymentsNum >= 1 &&
        oneTimePaymentsData.otherOneTimePayments.map((otherOneTimePayment, index) => (
          <ScheduledPayment
            key={index}
            scheduledPaymentDays={otherOneTimePayment.scheduledPaymentDays}
            scheduledPaymentDate={otherOneTimePayment.scheduledDate}
            paymentAmount={otherOneTimePayment.amount}
            isViewOnly={otherOneTimePayment.isViewOnly}
            isPaymentHistory={isPaymentHistory}
            accountNumber={otherOneTimePayment.accountNumber}
            paymentRowId={otherOneTimePayment.paymentRowId}
            achLoanNumber={achLoanNumber}
            onManagePayment={onManagePayment}
          />
        ))}
      {((!isAutoPay && !hasOneTimePayment) || !isCurrent) && isBilled && (
        <NextPaymentCard
          nextPaymentDays={nextPaymentData.nextPaymentDays}
          paymentAmount={nextPaymentData.paymentAmount}
          dueDate={nextPaymentData.dueDate}
          isCurrent={isCurrent}
          isOutlined
          isAutoPay={isAutoPay}
          isLoanOverviewScreen={isLoanOverviewScreen}
          onSchedulePayment={onSchedulePayment}
          onEnrollInAutoPay={onEnrollInAutoPay}
        />
      )}
    </>
  );
};

export default NextPaymentView;
