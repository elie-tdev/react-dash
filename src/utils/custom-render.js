import React from 'react';
import { I18nextProvider } from 'react-i18next';
import { render } from '@testing-library/react';
import { CssBaseline } from '@material-ui/core';
import { BrowserRouter as Router } from 'react-router-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { MockedProvider } from '@apollo/client/testing';

import { MenuContextProvider } from '@/contexts';
import i18n from '@/translations';
import { LightTheme } from '@/themes';

const wrapper = ({ mocks = [], menuContextProviderProps = {} }) => {
  const Wrapper = ({ children }) => {
    return (
      <Router>
        <MockedProvider mocks={mocks} addTypename={false}>
          <ThemeProvider theme={LightTheme}>
            <MenuContextProvider {...menuContextProviderProps}>
              <CssBaseline />
              <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
            </MenuContextProvider>
          </ThemeProvider>
        </MockedProvider>
      </Router>
    );
  };

  return Wrapper;
};

const customRender = (ui, { mocks, menuContextProviderProps, ...restOptions } = {}) =>
  render(ui, { wrapper: wrapper({ mocks, menuContextProviderProps }), ...restOptions });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
