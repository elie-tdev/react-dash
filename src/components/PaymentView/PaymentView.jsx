import { useState, useMemo } from 'react';
import dateformat from 'dateformat';
import TimeAgo from 'javascript-time-ago';

import ScheduledPayment from './ScheduledPayment';
import EnrollInAutoPay from './EnrollInAutoPay';
import NextPaymentCard from './NextPaymentCard';
import AutoPayCard from './AutoPayCard';

import { formatDollar } from '@/utils/format';
import { formatDateString } from '@/properties';

const PaymentView = ({
  loan,
  isAutoPay,
  dueDate,
  isCurrent,
  paymentAmount,
  elevation = 15,
  isOutlined = false,
  onSchedulePayment,
  onEnrollInAutoPay,
  isBilled,
}) => {
  const [isScheduled, _setIsScheduled] = useState(false);

  const timeAgo = useMemo(() => new TimeAgo('en-US'), []);
  const formattedDueDate = useMemo(() => dateformat(dueDate, formatDateString), [dueDate]);
  const memoizedPaymentAmount = useMemo(() => formatDollar(paymentAmount), [paymentAmount]);

  return (
    <>
      {isAutoPay && (
        <AutoPayCard
          paymentDate={loan.paymentDate}
          paymentValue={loan.paymentAmount}
          nextPaymentDays={loan.nextPaymentDays}
          paymentType={'autoPay'}
          dark
        />
      )}
      {isScheduled && (
        <ScheduledPayment
          scheduledPaymentDate={formattedDueDate}
          paymentDate={formattedDueDate}
          scheduledPaymentDays={timeAgo.format(new Date(dueDate))}
        />
      )}
      {isBilled && (
        <NextPaymentCard
          nextPaymentDays={timeAgo.format(new Date(dueDate))}
          paymentAmount={memoizedPaymentAmount}
          dueDate={formattedDueDate}
          isCurrent={isCurrent}
          isOutlined={isOutlined}
          elevation={elevation}
          onSchedulePayment={onSchedulePayment}
          onEnrollInAutoPay={onEnrollInAutoPay}
        />
      )}
      {!isAutoPay && <EnrollInAutoPay onEnrollInAutoPay={onEnrollInAutoPay} />}
    </>
  );
};

export default PaymentView;
