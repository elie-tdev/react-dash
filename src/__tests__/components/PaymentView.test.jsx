import React from 'react';
import TimeAgo from 'javascript-time-ago';

import { render } from '@/utils/custom-render';
import PaymentView from '@/components/PaymentView/PaymentView';

describe('PaymentView Component', () => {
  const loan = {
    paymentAmount: 500,
    paymentDate: 'Jan 17, 2021',
    nextPaymentDays: new TimeAgo('en-US').format(Date.now() + 3 * 24 * 60 * 60 * 1000),
  };
  test('renders PaymentView', () => {
    const mock = jest.fn();
    const { container } = render(
      <PaymentView
        loan={loan}
        paymentValue={500}
        dueDate="Jan 17, 2021"
        due={500}
        isOutlined
        onSchedulePayment={mock}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
