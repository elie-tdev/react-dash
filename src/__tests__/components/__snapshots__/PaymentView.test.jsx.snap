// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentView Component renders PaymentView 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-card-8 makeStyles-card-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root-34 makeStyles-box-2"
    >
      <h1
        class="MuiTypography-root makeStyles-title-3 MuiTypography-h1"
      >
        Enroll in AutoPay and never miss a payment again.
      </h1>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-root-35 makeStyles-root-36 MuiButton-text makeStyles-button-4 MuiButton-textSizeLarge MuiButton-sizeLarge MuiButton-disableElevation"
        data-testid="PrimaryButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Enroll in AutoPay
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
